Mrazadar/pon 12/subscription stripe post

mrazadar opened the pull request:
<!--
Please give the pull request a short but descriptive title.

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
-->

Subscription Checkout Post API Integration

- Integrated `stripe-checkout` api request
- Tested the success use-cases
- Tested the Transaction declined use-cases
- Handled the stripe API error states
- Validated error messages for empty and expired subscriptions 
- Fixed Subscription Alerts bug with Payment Alerts
- Fixed the Confirmation Modal broken tests
- Code cleanup

mrazadar committed:
feat: PON-12 subscription stripe post checkout api integration

feat: PON-12 implemented happy post checkout request

feat: PON-12 handle stripe api errors

feat: PON-12 validated error messages for empty and expired subscriptions

feat: PON-12 fixed Subscription alerts bug with Payment Alerts

feat: PON-12 fixed the ConfirmationModal broken test

refactor: PON-12 clean some code

refactor: PON-12 removed commented code

refactor: PON-12 removed expired subscritpion error handling

NawfalAhmed commented on src/subscription/__factories__/subscriptionStatus.factory.js:
```diff
@@ -0,0 +1,14 @@
+import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies
+
+Factory.define('subscriptionStatus')
+  // .sequence('subscription_id')
```
nit: maybe we could remove this commented code on line 4?

NawfalAhmed commented on src/subscription/subscription-methods/stripe/service.js:
```diff
@@ -51,61 +42,29 @@ export async function subscriptionStripeCheckout(
+    });
+    throw new Error((error?.message ? `${error?.name}: ${error.message}` : 'Something went wrong!'), {
+      cause: 'create-paymentMethod',
+      // cause: error.name,
```
nit: maybe we could remove this commented code on line 65?

NawfalAhmed commented on src/subscription/index.scss:
```diff
@@ -64,6 +64,9 @@
   }
 
+  .monthly-legal-notification p.micro {
+    font-size: 12px !important; 
```
Not sure, should we prefer `rem` here over `px`?

mrazadar replied:
We need to make it specific to 'px'  
Because it's a parallel text to stripe legal text and we want to make it look exactly same.

NawfalAhmed replied:
I see, thanks for the clarification

NawfalAhmed reviewed (approved):
Nice work, LGTM

julianajlk commented on src/subscription/confirmation-modal/ConfirmationModal.jsx:
```diff
@@ -30,19 +33,33 @@ const messages = defineMessages({
   const intl = useIntl();
-  // TODO: add the redirect URL logic for `Goto Dashboard` button
+  const { confirmationStatus } = useSelector(subscriptionStatusSelector);
+  const [isOpen, setOpen] = useState(false);
```
I didn't see a close button on the mockup, curious if a learner is supposed to be able to close the modal if they click outside of the modal?

mrazadar replied:
As discussed in today's standup we don't want to close it even clicked outside.

julianajlk commented on src/subscription/data/details/sagas.js:
```diff
@@ -91,31 +82,31 @@ export function* handleSubmitPayment({ payload }) {
-      // TODO: implement submitSubscriptionPayment errors with handleSubscriptionErrors
-      if (error.code) {
-        yield call(handleSubscriptionErrors, { messages: [error] }, true);
+      if (error.message && error.cause === 'create-paymentMethod') { // stripe payment method creation error
```
Good catch here.

julianajlk reviewed (approved):
Good work Raza! LGTM

mrazadar commented on src/subscription/data/status/reducer.js:
```diff
@@ -0,0 +1,42 @@
+export const subscriptionStatusReducer = (state = subscriptionStatusInitialState, action = null) => {
+  if (action !== null) {
+    switch (action.type) {
+      case submit3DS.TRIGGER: return { ...state, submitting: true };
```
I wanted to handle stripe status (including 3DS) in a separate redux node so just added some sample code which we can progress on while working on 3DS.

mrazadar committed:
feat: PON-125 adding segment events (#738)

feat: PON-125 added stripe payment form rendered event

feat: PON-125 added stripe submit clicked events

feat: PON-125 added subscription checkout page start paint event

feat: PON-125 added subscription submit success/failure events

refactor: PON-12 removed commented code
